### PR TITLE
Adds proof harness for aws_byte_buf_advance function

### PR DESCRIPTION
**Pull request by feliperodri**
*Description of changes:*
1. Updates implementation of `aws_byte_buf_advance` function with pre- and post-conditions;
2. Adds a proof harness for `aws_byte_buf_advance` function;

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by feliperodri** [comment]
This PR depends on the [#359](https://github.com/awslabs/aws-c-common/pull/359)

File: .cbmc-batch/jobs/aws_byte_buf_advance/Makefile [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+###########
+include ../Makefile.aws_byte_buf
+
+MAX_BUFFER_SIZE=40
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_advance/Makefile]: why 40?

File: .cbmc-batch/source/utils.c [hunk 2]
```diff
@@ -38,7 +38,7 @@ void assert_all_zeroes(const uint8_t *const a, const size_t len) {
 }
 
 void assert_byte_from_buffer_matches(const uint8_t *const buffer, const struct store_byte_from_buffer *const b) {
-    if (buffer) {
+    if (buffer && b) {
```
Comment (danielsn) [inline .cbmc-batch/source/utils.c]: What happens if b is null?

Comment (feliperodri) [reply .cbmc-batch/source/utils.c]: See https://github.com/awslabs/aws-c-common/pull/359#discussion_r286764364

**Comment by feliperodri** [comment]
@justinboswell @JonathanHenson @ColdenCullen @graebm could you review the changes in `byte_buf.h` and `byte_buf_test.c`?

File: tests/byte_buf_test.c [hunk 3]
```diff
@@ -233,11 +233,10 @@ static int s_test_buffer_advance(struct aws_allocator *allocator, void *ctx) {
     (void)allocator;
 
     uint8_t arr[16];
-    struct aws_byte_buf src_buf = aws_byte_buf_from_array(arr, sizeof(arr));
+    struct aws_byte_buf src_buf = aws_byte_buf_from_empty_array(arr, sizeof(arr));
 
-    struct aws_byte_buf dst_buf;
+    struct aws_byte_buf dst_buf = {.len = 0};
```
Comment (justinboswell) [inline tests/byte_buf_test.c]: Why not ```= {0}```?

Comment (feliperodri) [reply tests/byte_buf_test.c]: Done.

**Comment by feliperodri** [comment]
@tautschnig @karkhaz @angelhof @danielsn @markrtuttle could you review the proof harnesses?

**Pull request merged**